schema: allow transaction retry in schema changer backfill

ecwall opened the pull request:
Fixes #106696

This change adds a test to simulate transaction retries during the schema
changer backfill.

Some retries caused `ingested key collides with an existing one` errors
during addSSTable because the schema change's source data changed.
For example: `CREATE TABLE t AS SELECT nextval('seq');` - `nextval` is
non-transactional so each time it is selected, the value changes.

This change also allows the schema changer to specify that addSSTable is
allowed to shadow (replace) keys' values if they are different to handle
this case.

Release note: None

stevendanna committed:
kv,sql: testing knob to randomly inject retryable errors

Occasionally, we see bugs caused by mismangement of state inside a
function that will be retried. Namely, the functions `(*kv.DB).Txn`
and `(*sql.InternalExecutor).Txn` take funcs that must be suitably
idempotent.

This is testing knob provides cheap way to help us hunt down such
bugs.

In the long run, it would be nice to turn this on by default, but we
should probably make sure TestServer runs reliably with this turned on
before flipping the default.

I also considered retrying _every_ transaction once rather than
relying on randomness. However, that would require adding more state
in kv.Txn, which I wasn't sure I really wanted to do for testing
purposes.

Informs #106417

Epic: none

Release note: None

ecwall committed:
schema: allow transaction retry in schema changer backfill

Fixes #106696

This change adds a test to simulate transaction retries during the schema
changer backfill.

Some retries caused `ingested key collides with an existing one` errors
during addSSTable because the schema change's source data changed.
For example: `CREATE TABLE t AS SELECT nextval('seq');` - `nextval` is
non-transactional so each time it is selected, the value changes.

This change also allows the schema changer to specify that addSSTable is
allowed to shadow (replace) keys' values if they are different to handle
this case.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107954)
<!-- Reviewable:end -->

ecwall committed:
testing: replace EnableRandomTransactionRetryErrors with TransactionRetryFilter

Adds TransactionRetryFilter which allows the retry behavior to be customized.

EnableRandomTransactionRetryErrors has been refactored to a
TransactionRetryFilter implementation that randomly retries.

Release note: None

fqazi reviewed (commented):
> Reviewed 9 of 9 files at r1, 13 of 13 files at r2, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @rytaft, and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/testing_knobs.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107954#-NbUpU_P9o5NDnd_bDXG:-NbUpU_P9o5NDnd_bDXH:b-d1t8gn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3069d1d2360e656656c4fb3ae8f82b32050ba63a/pkg/kv/kvclient/kvcoord/testing_knobs.go#L64)):*
> ```Go
> 	// loops to randomly inject retriable errors.
> 	EnableRandomTransactionRetryErrors bool
> }
> ```

Should this setting be a probability value instead?
___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107954#-NbUppnM-nn0qoY-2S9Y:-NbUppnM-nn0qoY-2S9Z:behj9em) ([raw file](https://github.com/cockroachdb/cockroach/blob/3069d1d2360e656656c4fb3ae8f82b32050ba63a/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L43)):*
> ```Go
> //
> // See: https://github.com/cockroachdb/cockroach/pull/73512.
> var DisableCommitSanityCheck = envutil.EnvOrDefaultBool("COCKROACH_DISABLE_COMMIT_SANITY_CHECK", false)
> ```

Should this be in this commit?
___
*[`pkg/sql/rowexec/bulk_row_writer.go` line 146 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107954#-NbUqcyyEo8y8LkHTXDi:-NbUqcyyEo8y8LkHTXDj:b2v2mrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/41820281ec982b6aeb4a542ed7dda11255097309/pkg/sql/rowexec/bulk_row_writer.go#L146)):*
> ```Go
> 			Name:          sp.tableDesc.GetName(),
> 			MinBufferSize: bufferSize,
> 			// We disallow shadowing here to ensure that we report errors when builds
> ```

Update the comment here, also I think this will prevent errors from surfacing I think in other cases?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107954#-NbUppnM-nn0qoY-2S9Y:-NbZe3af5Fa9liVoX3rt:b-mz3oav) ([raw file](https://github.com/cockroachdb/cockroach/blob/3069d1d2360e656656c4fb3ae8f82b32050ba63a/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L43)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Should this be in this commit?
</blockquote></details>

This is a change from @stevendanna that has not been merged yet. My change extends some of the functionality.
___
*[`pkg/sql/rowexec/bulk_row_writer.go` line 146 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107954#-NbUqcyyEo8y8LkHTXDi:-NbZeHR456zw4BCCX6az:bwmaaye) ([raw file](https://github.com/cockroachdb/cockroach/blob/41820281ec982b6aeb4a542ed7dda11255097309/pkg/sql/rowexec/bulk_row_writer.go#L146)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Update the comment here, also I think this will prevent errors from surfacing I think in other cases?
</blockquote></details>

After talking with the DR team (https://cockroachlabs.slack.com/archives/C03JCUUSCD6/p1691669801528989), I will have to rethink this part of the PR.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/testing_knobs.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107954#-NbUpU_P9o5NDnd_bDXG:-NbZebqr25TKr6-Y_8Ee:b-abo6mc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3069d1d2360e656656c4fb3ae8f82b32050ba63a/pkg/kv/kvclient/kvcoord/testing_knobs.go#L64)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Should this setting be a probability value instead?
</blockquote></details>

Only some of the filtering is probability based so I left it to the filter impl to do that calculation.


<!-- Sent from Reviewable.io -->

ecwall committed:
testing: add comments to create_as_test.go

Release note: None

ecwall committed:
testing: replace StartNewTestCluster with StartServer in create_as_test.go

Release note: None

fqazi commented on pkg/sql/rowexec/bulk_row_writer.go:
```diff
@@ -133,6 +133,21 @@ func (sp *bulkRowWriter) wrapDupError(ctx context.Context, orig error) error {
 	writeTS := sp.spec.Table.CreateAsOfTime
+
+	// Delete existing span before ingestion to prevent key collisions.
+	if sp.spec.DeleteSpan {
```
If this operation is distributed, don't we risk stepping on ourselves? Should this be done at a higher level?

dt replied:
Yeah, I don't think we should be doing this in the processor; I think we should be doing this in the coordinator before we setup the flow, since a) now every proc is deleting the same span, which will all the levels of the LSM with duplicate range keys pushing all writes into L0, and more importantly b) one proc might already be writing by the time another starts up, gets scheduled and runs its deletes, leading to data corruption.

fqazi reviewed (approved):
> Reviewed 18 of 18 files at r9, 13 of 13 files at r10, 1 of 1 files at r11, 1 of 1 files at r12, 8 of 8 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @stevendanna)



<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=fqazi
